{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import shutil\n",
    "import pysam\n",
    "import reportlab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "source_path = '/Users/yijiasun77/Desktop/Intern/Rayca/normal_sample.deepvariant.vcf.gz'\n",
    "destination_path = '/Users/yijiasun77/Desktop/Intern/Rayca/normal_sample.deepvariant.vcf'\n",
    "\n",
    "with gzip.open(source_path, 'rb') as f_in:\n",
    "    with open(destination_path, 'wb') as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "from reportlab.lib.pagesizes import letter\n",
    "from reportlab.pdfgen import canvas\n",
    "from reportlab.lib import colors\n",
    "from reportlab.lib.styles import getSampleStyleSheet\n",
    "from reportlab.platypus import SimpleDocTemplate, Table, TableStyle, Paragraph, Spacer\n",
    "\n",
    "def parse_vcf(vcf_file):\n",
    "    vcf = pysam.VariantFile(vcf_file)\n",
    "    variants = []\n",
    "    for rec in vcf.fetch():\n",
    "        for alt in rec.alts:\n",
    "            variants.append({\n",
    "                'chromosome': rec.chrom,\n",
    "                'position': rec.pos,\n",
    "                'ref_allele': rec.ref,\n",
    "                'alt_allele': alt,\n",
    "                'quality': rec.qual,\n",
    "                'filter': rec.filter.keys(),\n",
    "                'info': rec.info\n",
    "                \n",
    "            })\n",
    "    return variants\n",
    "\n",
    "\n",
    "def classify_variant(variant):\n",
    "    return 'Pathogenic' if variant['quality'] > 30 else 'Likely Pathogenic'\n",
    "\n",
    "\n",
    "def generate_pdf_report(variants, pdf_filename):\n",
    "    doc = SimpleDocTemplate(pdf_filename, pagesize=letter)\n",
    "    story = []\n",
    "    styles = getSampleStyleSheet()\n",
    "    \n",
    "    for variant in variants:   \n",
    "\n",
    "        story.append(Paragraph('<b>Patient ID:</b> {}'.format(patient_info['Patient ID']), styles['Normal']))\n",
    "        story.append(Paragraph('<b>Date of Birth:</b> {}'.format(patient_info['Date of Birth']), styles['Normal']))\n",
    "        story.append(Paragraph('<b>Gender:</b> {}'.format(patient_info['Gender']), styles['Normal']))\n",
    "        story.append(Paragraph('<b>Ethnicity:</b> {}'.format(patient_info['Ethnicity']), styles['Normal']))\n",
    "        story.append(Paragraph('<b>Family History:</b> {}'.format(patient_info['Family History']), styles['Normal']))\n",
    "        story.append(Paragraph(f\"Chromosome: {variant['chromosome']}\", styles['Normal']))\n",
    "        story.append(Paragraph(f\"Position: {variant['position']}\", styles['Normal']))\n",
    "        story.append(Paragraph(f\"Reference Allele: {variant['ref_allele']}\", styles['Normal']))\n",
    "        story.append(Paragraph(f\"Alternate Allele: {variant['alt_allele']}\", styles['Normal']))\n",
    "        story.append(Paragraph(f\"Quality: {variant['quality']}\", styles['Normal']))\n",
    "        story.append(Paragraph(f\"Filter: {', '.join(variant['filter'])}\", styles['Normal']))\n",
    "        \n",
    "        \n",
    "        classification = classify_variant(variant)\n",
    "        story.append(Paragraph(f\"ACMG Classification: {classification}\", styles['Normal']))\n",
    "        \n",
    "        story.append(Spacer(1, 12))\n",
    "    \n",
    "    doc.build(story)\n",
    "\n",
    "\n",
    "\n",
    "vcf_file = 'normal_sample.deepvariant.vcf'  # Path to the VCF file\n",
    "parsed_variants = parse_vcf(vcf_file)\n",
    "\n",
    "\n",
    "patient_info = {\n",
    "    'Patient ID': '[Insert Patient ID]',\n",
    "    'Date of Birth': '[Insert Date of Birth]',\n",
    "    'Gender': '[Insert Gender]',\n",
    "    'Ethnicity': '[Insert Ethnicity]',\n",
    "    'Family History': '[Insert relevant family history information]'\n",
    "}\n",
    "\n",
    "# Classify each variant\n",
    "for variant in parsed_variants:\n",
    "    variant['classification'] = classify_variant(variant)\n",
    "\n",
    "# Generate the report\n",
    "pdf_filename = 'Rayca_clinical_report-Yijia Sun.pdf'\n",
    "generate_pdf_report(parsed_variants, pdf_filename)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
